File: apps/holefeeder-mobile/src/flows/core/flows/cashflow.ts
import { CategoryType } from '@/flows/core/categories/category-type';
import { DateIntervalType } from '@/shared/core/date-interval-type';
import { DateOnly } from '@/shared/core/date-only';
import { Id } from '@/shared/core/id';
import { Money } from '@/shared/core/money';
import { Result } from '@/shared/core/result';
import { Validate, Validator } from '@/shared/core/validate';
import { TagList } from './tag-list';

export type Cashflow = {
  id: Id;
  effectiveDate: DateOnly;
  amount: Money;
  intervalType: DateIntervalType;
  frequency: number;
  recurrence: number;
  description: string;
  accountId: Id;
  categoryId: Id;
  categoryType: CategoryType;
  inactive: boolean;
  tags: TagList;
};

export const CashflowErrors = {
  invalid: 'cashflow-invalid',
  neverPaid: 'cashflow-never-paid',
};

const isPositiveNumber = Validator.number({ min: 1 });
const isValidBoolean = Validator.boolean();

const create = (value: Record<string, unknown>): Result<Cashflow> =>
  Result.combine<Cashflow>({
    id: Id.create(value.id),
    effectiveDate: DateOnly.create(value.effectiveDate),
    amount: Money.create(value.amount),
    intervalType: DateIntervalType.create(value.intervalType),
    frequency: Validate.validate(isPositiveNumber, value.frequency, [CashflowErrors.invalid]),
    recurrence: Validate.validate(isPositiveNumber, value.recurrence, [CashflowErrors.invalid]),
    description: Result.success(value.description as string),
    accountId: Id.create(value.accountId),
    categoryId: Id.create(value.categoryId),
    categoryType: CategoryType.create(value.categoryType),
    inactive: Validate.validate(isValidBoolean, value.inactive, [CashflowErrors.invalid]),
    tags: TagList.create(value.tags),
  });

const valid = (value: Record<string, unknown>): Cashflow => ({
  id: Id.valid(value.id),
  effectiveDate: DateOnly.valid(value.effectiveDate),
  amount: Money.valid(value.amount),
  intervalType: DateIntervalType.valid(value.intervalType),
  frequency: value.frequency as number,
  recurrence: value.recurrence as number,
  description: value.description as string,
  accountId: Id.valid(value.accountId),
  categoryId: Id.valid(value.categoryId),
  categoryType: CategoryType.valid(value.categoryType),
  inactive: value.inactive as boolean,
  tags: TagList.valid(value.tags),
});

export const Cashflow = {
  create: create,
  valid: valid,
};

File: backend/src/Holefeeder.Domain/Features/Transactions/Cashflow.cs
using Holefeeder.Domain.Enumerations;
using Holefeeder.Domain.Features.Accounts;
using Holefeeder.Domain.Features.Categories;
using Holefeeder.Domain.Features.Users;
using Holefeeder.Domain.ValueObjects;

namespace Holefeeder.Domain.Features.Transactions;

public partial record Cashflow : IAggregateRoot
{
    private IReadOnlyCollection<string> _tags = [];

    private Cashflow(CashflowId id, DateOnly effectiveDate, DateIntervalType intervalType, AccountId accountId,
        CategoryId categoryId, UserId userId)
    {
        Id = id;
        EffectiveDate = effectiveDate;
        IntervalType = intervalType;
        AccountId = accountId;
        CategoryId = categoryId;
        UserId = userId;
    }

    public CashflowId Id { get; }

    public DateOnly EffectiveDate { get; private init; }

    public DateIntervalType IntervalType { get; private init; }

    public int Frequency { get; private init; }

    public int Recurrence { get; private init; }

    public Money Amount { get; private init; }

    public string Description { get; private init; } = string.Empty;

    public AccountId AccountId { get; private init; }

    public Account? Account { get; private init; }

    public CategoryId CategoryId { get; private init; }

    public Category? Category { get; init; }

    public IReadOnlyCollection<string> Tags => _tags.ToImmutableList();

    public IReadOnlyCollection<Transaction> Transactions { get; } = new List<Transaction>();

    public DateOnly? LastPaidDate => Transactions.Count > 0 ? Transactions.Max(x => x.Date) : null;

    public DateOnly? LastCashflowDate => Transactions.Max(x => x.CashflowDate);

    public bool Inactive { get; private init; }

    public UserId UserId { get; private init; }
}

public sealed record CashflowId : StronglyTypedId<CashflowId>;
